common/lockmanager_test.go

```diff
@@ -0,0 +1,318 @@
+        test.Fatalf("Failed to return an error when unlocking a lock that is already unlocked.");
+    }
+}
+
+func TestUnlockingKeyThatDoesntExist(test *testing.T) {
+    doesNotExistKey := "dne";
+
+    err := Unlock(doesNotExistKey);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a key that does not exist.");
+    }
+}
+
+func TestLockUnlockDifferentKeysOneThread(test *testing.T) {
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var verifyWaitGroup sync.WaitGroup;
+    var threadWaitGroup sync.WaitGroup;
+
+    // Lock key1 for the first time.
+    Lock(key1);
+
+    threadWaitGroup.Add(1);
+    verifyWaitGroup.Add(1);
+
+    // This thread should lock key2 for the first time while key1 is already locked.
+    go func() {
+        defer threadWaitGroup.Done();
+        Lock(key2);
+        verifyWaitGroup.Done();
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock.");
+        }
+    }()
+
+    // Make sure the thread locked key2 before the main thread unlocks key1.
+    verifyWaitGroup.Wait();
+
+    // Unlock key1.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    // Wait for the thread to finish unlocking key2.
+    threadWaitGroup.Wait();
+}
+
+func TestLockUnlockDifferentKeysTwoThreads(test *testing.T) {
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var threadsWaitGroup sync.WaitGroup;
+    var funcWaitGroup sync.WaitGroup
+
+    threadsWaitGroup.Add(2);
+    funcWaitGroup.Add(2);
+
+    // First thread locks and unlocks key1.
+    go func() {
+        defer funcWaitGroup.Done();
+        Lock(key1);
+        threadsWaitGroup.Done();
+        err := Unlock(key1);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock.")
+        }
+    }()
+
+    // Second thread locks and unlocks key2.
+    go func() {
+        defer funcWaitGroup.Done();
+        Lock(key2);
+        threadsWaitGroup.Done();
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to Unlock")
+        }
+    }()
+    
+    // Wait until both threads have finished locking their keys so they 
+    // can unlock at the same time.
+    threadsWaitGroup.Wait();
+    
+    // Wait until both threads finish their execution.
+    funcWaitGroup.Wait();
+}
+
+func TestLockingTwiceWithSameKeyConcurrently(test *testing.T) {
+    key1 := "testkey1";
+    var firstThreadWaitGroup sync.WaitGroup;
+    var secondThreadWaitGroup sync.WaitGroup;
+    var bothThreadsWaitGroup sync.WaitGroup;
+
+    firstThreadWaitGroup.Add(1);
+    secondThreadWaitGroup.Add(1);
+    bothThreadsWaitGroup.Add(1);
+
+    // First thread aquires the lock a first time.
+    go func() {
+        defer firstThreadWaitGroup.Done();
+        Lock(key1);
+    }()
+
+    // Wait for the first thread to lock key1. 
+    firstThreadWaitGroup.Wait();
+
+
+    // Second thread aquires the lock a second time.
```

Comment:
This comment is a little misleading.
You want to say that the second thread tries to acquire the lock, or that is blocks to acquire the lock.
We don't want the reader thinking that it was a direct acquisition like the first routine.